### PR TITLE
PHP 7.4.0

### PR DESCRIPTION
**Pull request by alcaeus**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

Supersedes #47288. I took the base formula from @javian a while back and I've been running it locally for a while now.

### CODE REVIEW

**Comment by alcaeus** [comment]
@SMillerDev I'm seeing build issues due to some ldap symbols missing:
> Undefined symbols for architecture x86_64:
  "_ldap_create_passwordpolicy_control", referenced from:
      _zif_ldap_exop_passwd in ldap.o

I've removed some `ENV` settings that were present before, let's hope that passes. I'm a bit at a loss since I got this to compile on my Catalina machine 🤔

**Comment by jaikdean** [comment]
I get the same error as the CI build when building locally on macOS Catalina 10.15.1 with `brew upgrade https://raw.githubusercontent.com/Homebrew/homebrew-core/6998ab433bc03d3dd13a8e141673315a26f5ed03/Formula/php.rb`.

There's a related bug report from 7.3.0 that's to do with the macOS version of OpenLDAP being different to the open source one. I know next to nothing about OpenLDAP unfortunately, so it's not really clear to me what's going on.

When you built it successfully on your machine, did you have OpenLDAP installed via Homebrew? On my machine that failed I do:

```
$ brew info openldap
openldap: stable 2.4.48 (bottled) [keg-only]
```

**Comment by SMillerDev** [comment]
https://github.com/SMillerDev/homebrew-core/commit/50a2e97f48ed94e77588764e729a744d8c444052 This is my personal set of solutions from a while ago. I seem to have disabled LDAP all together.

**Comment by Bo98** [comment]
It's likely linking against system libraries rather than Homebrew's OpenLDAP, something that is supposed to be fixed by the patch that was there for PHP 7.3, but it seems that the new one might not be working as expected. You can check this by downloading the generated bottles and running `otool` (or install them and use `brew linkage`).

File: Formula/php.rb [hunk 1]
```diff
@@ -417,65 +419,44 @@ def plist; <<~EOS
...
-index 168c465f8d..6c087d152f 100644
---- a/acinclude.m4
-+++ b/acinclude.m4
-@@ -441,7 +441,11 @@ dnl
- dnl Adds a path to linkpath/runpath (LDFLAGS)
+diff --git a/build/php.m4 b/build/php.m4
+index 3624a33a8e..d17a635c2c 100644
+--- a/build/php.m4
++++ b/build/php.m4
+@@ -423,7 +423,7 @@ dnl
```
Comment (derrabus) [inline Formula/php.rb]:
```suggestion
@@ -425,7 +425,7 @@ dnl
```

🤓

File: Formula/php.rb [hunk 2]
```diff
@@ -417,65 +419,44 @@ def plist; <<~EOS
...
--    ])
--  fi
-+    ]) ;;
-+  esac
- ])
-
- dnl
-@@ -487,7 +491,11 @@ dnl add an include path.
- dnl if before is 1, add in the beginning of INCLUDES.
+@@ -468,7 +468,7 @@ dnl
```
Comment (derrabus) [inline Formula/php.rb]:
```suggestion
@@ -470,7 +470,7 @@ dnl
```

File: Formula/php.rb [hunk 3]
```diff
@@ -81,6 +83,10 @@ def install
     # Prevent homebrew from harcoding path to sed shim in phpize script
     ENV["lt_cv_path_SED"] = "sed"
 
+    # system pkg-config missing
+    ENV["KERBEROS_CFLAGS"] = " "
+    ENV["KERBEROS_LIBS"] = "-lkrb5"
```
Comment (derrabus) [inline Formula/php.rb]:
```suggestion
    ENV["KERBEROS_LIBS"] = "-lkrb5"
    ENV["SASL_CFLAGS"] = "-I#{MacOS.sdk_path_if_needed}/usr/include/sasl"
    ENV["SASL_LIBS"] = "-lsasl2"
    ENV["EDIT_CFLAGS"] = " "
    ENV["EDIT_LIBS"] = "-ledit"
```

I needed some more `ENV` variables to get this formula to work.

**Review by derrabus — CHANGES_REQUESTED** [review]
With the added `ENV` variables, the formula builds on my Catalina MBP.

**Comment by pascaldevink** [comment]
Without these `ENV` variables, I'm getting this error about libsasl2 when using `brew upgrade https://raw.githubusercontent.com/Homebrew/homebrew-core/39551efd21841d0069f374c3e913b6b8af5a969e/Formula/php.rb` on macOS 10.15.1:
```
checking for libsasl2... no
configure: error: Package requirements (libsasl2) were not met:

No package 'libsasl2' found
```

**Comment by Bo98** [comment]
There's a number of changes to the configure options that should be reflected here:

* `--enable-opcache-file` no longer exists. It is now an INI option.
* `--enable-zip` has been renamed `--with-zip`.
* `--with-curl` no longer accepts a directory. Use `--with-curl` on its own and `pkg-config` will find Homebrew's curl.
* `--with-freetype-dir` no longer exists. Use `--with-freetype` on its own (no directory passed).
* `--with-gd` no longer exists. Use `--enable-gd` to enable the extension and `--with-external-gd` to use Homebrew's `gd` rather than the bundled one (unless Homebrew's one is incompatible).
* `--with-icu-dir` no longer exists. There is no replacement - the dependency is mandatory when `--enable-intl` is passed and `pkg-config` is used to find it.
* `--with-jpeg-dir` no longer exists. Use `--with-jpeg` (on its own - no directory passed).
* `--with-kerberos` no longer accepts a directory. Use `--with-kerberos` on its own (and the necessary ENV variables since we don't have a system pkg-config for it at the moment).
* `--with-ldap-sasl` no longer accepts a directory. Use `--with-ldap-sasl` on its own (and the necessary ENV variables since we don't have a system pkg-config for it at the moment).
* `--with-libxml-dir` no longer exists. Use `--with-libxml` (on its own - no directory passed).
* `--with-libedit` no longer accepts a directory. Use `--with-libedit` on its own (and the necessary ENV variables since we don't have a system pkg-config for it at the moment).
* `--with-libzip` no longer exists. This is covered by `--with-zip` now via `pkg-config`.
* `--with-openssl` no longer accepts a directory. Use `--with-openssl` on its own (no directory passed).
* `--with-pdo-sqlite` no longer accepts a directory. Use `--with-pdo-sqlite` on its own (no directory passed).
* `--with-png-dir` no longer exists. It is a mandatory dependency and is found via `pkg-config`.
* `--with-sodium` no longer accepts a directory. Use `--with-sodium` on its own (no directory passed).
* `--with-sqlite3` no longer accepts a directory. Use `--with-sqlite3` on its own (no directory passed).
* `--with-unixODBC` can now be used without a directory. I recommend doing so, but it can still be passed a directory if there's problems.
* `--with-webp-dir` no longer exists. Use `--with-webp` (on its own - no directory passed).
* `--with-xsl` no longer accepts a directory. Use `--with-xsl` on its own (no directory passed).
* `--with-zlib` no longer accepts a directory. Use `--with-zlib` on its own (no directory passed).

**Comment by alcaeus** [comment]
I’ll take care of issues later - I was conservative in throwing out options since it built in my machine, but I’ll just add the whole set and go from there.

**Comment by Bo98** [comment]
Most probably don't matter - but the ones that have been renamed certainly will as some features may now be disabled as a result.

**Comment by alcaeus** [comment]
Thanks for that excellent list! I've applied all changes as I had them in my own formula, so hopefully we'll get it to compile now. I'll check back once the build has completed and work on any remaining issues then.

**Comment by alcaeus** [comment]
Build stable! 🎉

**Comment by pascaldevink** [comment]
When building the latest version, I'm getting this error about iconv:
```
==> make
Last 15 lines from /Users/pascal/Library/Logs/Homebrew/php/03.make:
make: *** [sapi/cli/php] Error 1
      _php_iconv_string in iconv.o
      __php_iconv_strlen in iconv.o
      _zif_iconv_substr in iconv.o
      __php_iconv_strpos in iconv.o
      _zif_iconv_mime_encode in iconv.o
      __php_iconv_mime_decode in iconv.o
      _php_iconv_stream_filter_factory_create in iconv.o
      ...
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [libs/libphp7.bundle] Error 1
warning: no debug symbols in executable (-arch x86_64)
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: file: ext/opcache/.libs/opcache.a(shared_alloc_shm.o) has no symbols
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: file: ext/opcache/.libs/opcache.a(shared_alloc_shm.o) has no symbols
```

libiconv  is installed:
```
$ brew info libiconv
libiconv: stable 1.16 (bottled) [keg-only]
```

Uninstalling libiconv made it compile though, but the error can throw people off. I'm not sure if there's a way around it?

**Comment by alcaeus** [comment]
@pascaldevink I installed the `libiconv` extension:
```
$ brew info libiconv
libiconv: stable 1.16 (bottled) [keg-only]
Conversion library
https://www.gnu.org/software/libiconv/
/usr/local/Cellar/libiconv/1.16 (30 files, 2.4MB)
  Poured from bottle on 2019-11-29 at 09:02:31
From: https://github.com/Homebrew/homebrew-core/blob/master/Formula/libiconv.rb
==> Caveats
libiconv is keg-only, which means it was not symlinked into /usr/local,
because macOS already provides this software and installing another version in
parallel can cause all kinds of trouble.

If you need to have libiconv first in your PATH run:
  echo 'export PATH="/usr/local/opt/libiconv/bin:$PATH"' >> ~/.bash_profile

For compilers to find libiconv you may need to set:
  export LDFLAGS="-L/usr/local/opt/libiconv/lib"
  export CPPFLAGS="-I/usr/local/opt/libiconv/include"
```

Then, I reinstalled the formula:
```
alcaeus at vulcan in ~/Code/homebrew/homebrew-core on php-7.4.0
$ brew reinstall --build-from-source Formula/php.rb
==> Reinstalling php 
Warning: A newer Command Line Tools release is available.
Update them from Software Update in System Preferences.

==> Downloading https://www.php.net/distributions/php-7.4.0.tar.xz
Already downloaded: /Users/alcaeus/Library/Caches/Homebrew/downloads/e77e2ae71ff7dcc4a0f5e4b4fa946e9bc4c610b00a6a3df0e67e53f2cd01927c--php-7.4.0.tar.xz
==> Patching
patching file build/php.m4
patching file configure.ac
==> ./buildconf --force
==> ./configure --prefix=/usr/local/Cellar/php/7.4.0 --localstatedir=/usr/local/var --sysconfdir=/usr/local/etc/php/7.4 --with-config-file-path=/usr/local/etc/php/7.4 --with-config-file-scan-dir=/usr/loca
==> make
==> make install
==> /usr/local/Cellar/php/7.4.0/bin/pear config-set php_ini /usr/local/etc/php/7.4/php.ini system
==> /usr/local/Cellar/php/7.4.0/bin/pear config-set php_dir /usr/local/share/pear system
==> /usr/local/Cellar/php/7.4.0/bin/pear config-set doc_dir /usr/local/share/pear/doc system
==> /usr/local/Cellar/php/7.4.0/bin/pear config-set ext_dir /usr/local/lib/php/pecl/20190902 system
==> /usr/local/Cellar/php/7.4.0/bin/pear config-set bin_dir /usr/local/opt/php/bin system
==> /usr/local/Cellar/php/7.4.0/bin/pear config-set data_dir /usr/local/share/pear/data system
==> /usr/local/Cellar/php/7.4.0/bin/pear config-set cfg_dir /usr/local/share/pear/cfg system
==> /usr/local/Cellar/php/7.4.0/bin/pear config-set www_dir /usr/local/share/pear/htdocs system
==> /usr/local/Cellar/php/7.4.0/bin/pear config-set man_dir /usr/local/share/man system
==> /usr/local/Cellar/php/7.4.0/bin/pear config-set test_dir /usr/local/share/pear/test system
==> /usr/local/Cellar/php/7.4.0/bin/pear config-set php_bin /usr/local/opt/php/bin/php system
==> /usr/local/Cellar/php/7.4.0/bin/pear update-channels
==> Caveats
[...]
==> Summary
🍺  /usr/local/Cellar/php/7.4.0: 512 files, 75.9MB, built in 13 minutes 42 seconds

alcaeus at vulcan in ~/Code/homebrew/homebrew-core on php-7.4.0
$ brew test Formula/php.rb
Testing php
==> /usr/local/Cellar/php/7.4.0/bin/php -i
==> /usr/local/Cellar/php/7.4.0/sbin/php-fpm -t
==> /usr/local/Cellar/php/7.4.0/bin/phpdbg -V
==> /usr/local/Cellar/php/7.4.0/bin/php-cgi -m
==> /usr/local/Cellar/php/7.4.0/bin/php -m
==> curl -s 127.0.0.1:62476
[29-Nov-2019 09:18:31] NOTICE: fpm is running, pid 70131
[29-Nov-2019 09:18:31] NOTICE: ready to handle connections
==> curl -s 127.0.0.1:62476
[29-Nov-2019 09:18:34] NOTICE: Terminating ...
[29-Nov-2019 09:18:34] NOTICE: exiting, bye-bye!
```

I can't reproduce the failure, so I'm a bit at a loss. Which version of MacOS are you running, and could you give it another shot with libiconv?

**Comment by pascaldevink** [comment]
@alcaeus reinstalling libiconv indeed seems to work 🎉

File: Formula/php.rb [hunk 4]
```diff
@@ -20,15 +20,18 @@ class Php < Formula
   depends_on "curl-openssl"
   depends_on "freetds"
   depends_on "freetype"
+  depends_on "gd"
```
Comment (SMillerDev) [inline Formula/php.rb]: Isn't this part of macOS? Why is this needed if we had the option enabled before?

File: Formula/php.rb [hunk 5]
```diff
@@ -20,15 +20,18 @@ class Php < Formula
...
   depends_on "glib"
   depends_on "gmp"
   depends_on "icu4c"
   depends_on "jpeg"
+  depends_on "libffi"
   depends_on "libpng"
   depends_on "libpq"
   depends_on "libsodium"
   depends_on "libzip"
+  depends_on "oniguruma"
```
Comment (SMillerDev) [inline Formula/php.rb]: This should be bundled by PHP

Comment (alcaeus) [reply Formula/php.rb]: Build fails without:

> checking for oniguruma... no
configure: error: Package requirements (oniguruma) were not met:
>
> No package 'oniguruma' found
>
> Consider adjusting the PKG_CONFIG_PATH environment variable if you
installed software in a non-standard prefix.
>
> Alternatively, you may set the environment variables ONIG_CFLAGS
and ONIG_LIBS to avoid the need to call pkg-config.
See the pkg-config man page for more details.

Given that, I'd keep it in.

File: Formula/php.rb (hunk 4 above)
Comment (alcaeus) [reply Formula/php.rb]: To quote:

> `--with-gd` no longer exists. Use `--enable-gd` to enable the extension and `--with-external-gd` to use Homebrew's gd rather than the bundled one (unless Homebrew's one is incompatible).

I've made a small mistake previously where I used `--with-gd` instead of `--with-external-gd`. I've fixed and tested locally and everything builds successfully. I can push that change, but I can also drop the dependency and use the bundled one instead.

Comment (SMillerDev) [reply Formula/php.rb]: I'd say the bundled one might cause less surprise bugs. Let's do that one.

Comment (Bo98) [reply Formula/php.rb]: Yeah both the bundled and external work. If the bundled version is linked statically then you can use either or - not sure if there's a preference PHP has or not. I think we don't use external `pcre` so feel free to drop the option.

Also note: the option also doesn't take in a directory so you don't need to do that.

File: Formula/php.rb (hunk 5 above)
Comment (Bo98) [reply Formula/php.rb]: Indeed Oniguruma is no longer bundled.

**Comment by Bo98** [comment]
> is there a way to prevent `brew` from truncating the configure call in the build output? 

~/Library/Logs/Homebrew/php

File: Formula/php.rb [hunk 6]
```diff
@@ -113,50 +122,47 @@ def install
...
       --with-bz2#{headers_path}
-      --with-curl=#{Formula["curl-openssl"].opt_prefix}
+      --with-curl
+      --with-ffi
       --with-fpm-user=_www
       --with-fpm-group=_www
-      --with-freetype-dir=#{Formula["freetype"].opt_prefix}
-      --with-gd
+      --with-freetype
+      --with-gd=#{Formula["gd"].opt_prefix}
```
Comment (derrabus) [inline Formula/php.rb]:
```suggestion
```

`--with-gd` is gone.

**Review by derrabus — CHANGES_REQUESTED** [review]

File: Formula/php.rb (hunk 4 above)
Comment (derrabus) [reply Formula/php.rb]: If we use the bundled GD, we need to specify which features it should support.

```
  --enable-gd             Include GD support
  --with-external-gd      Use external libgd
  --with-webp             GD: Enable WEBP support (only for bundled libgd)
  --with-jpeg             GD: Enable JPEG support (only for bundled libgd)
  --with-xpm              GD: Enable XPM support (only for bundled libgd)
  --with-freetype         GD: Enable FreeType 2 support (only for bundled
                          libgd)
  --enable-gd-jis-conv    GD: Enable JIS-mapped Japanese font support (only
                          for bundled libgd)
```

File: Formula/php.rb (hunk 6 above)
Comment (mrfu2) [reply Formula/php.rb]: As of php/php-src@570d431, it's renamed to `--enable-gd`. Additionally, there are some more changes to the GD-related flags: php/php-src@48ca2c0. [Source](https://github.com/phpbrew/phpbrew/issues/1020)

File: Formula/php.rb (hunk 4 above)
Comment (alcaeus) [reply Formula/php.rb]: I switched to bundled gd to maintain the status quo. Switching to external gd is best taken care of in a separate PR if maintainers agree that it should be done.

**Comment by Bo98** [comment]
Awesome. We all happy with this?

**Review by derrabus — APPROVED** [review]

**Comment by Bo98** [comment]
Thanks @alcaeus and everyone who commented.

`php@7.3` will follow within the next hour if all goes well.

**Comment by ravenberg** [comment]
Thank you

**Comment by adrianbj** [comment]
Thanks everyone!
